sailor/utils/utils.py

```diff
@@ -16,3 +18,15 @@ def _is_non_string_iterable(obj):
     if issubclass(obj.__class__, str):
         return False
     return isinstance(obj, Iterable)
+
+
+class WarningAdapter(logging.LoggerAdapter):
+    """Allow a logger to convert warnings logs into real warnings to simplify logging setup for users."""
+
+    def __init__(self, logger, extra=None):
+        super().__init__(logger, extra or {})
+
+    def log_with_warning(self, msg, warning_stacklevel=1, warning_category=None, *args, **kwargs):
+        """Delegate a warning call to the underlying logger and trigger a real warning with the same message."""
+        warnings.warn(msg, category=warning_category, stacklevel=warning_stacklevel + 1)
+        self.log(logging.WARNING, msg, *args, **kwargs)
```

Comment:
the `log` function also has a `stacklevel` arg that needs to be set to `warning_stacklevel + 1`.

If you don't do this, the logger will always log the message as it would be coming from the `log_with_warning` function.
You can see this if you configure your logger like this.

```
formatter = logging.Formatter("%(name)s %(funcName)s %(lineno)d [%(levelname)s] %(message)s")
handler = logging.StreamHandler()
handler.setFormatter(formatter)
logger.addHandler(handler)
```

Reply:
hm, not quite sure about this. so far we haven't (ever?) set a stack-level for logging. furthermore `warning`-logs should not behave differently than any other logs. So I'd argue the appropriate stack-level here is not necessarily the same as the `warning_stacklevel`.

I do agree that `log_stacklevel` should be at least `1`, to make the log appear in the original function. Furthermore, we could add a new argument `log_stacklevel` if we want to set a stack level for logging, but for now that argument would be unused, no?

Reply:
This is the normal behaviour, for, e.g. a regular `LOG.error('test')`
E.g.
```
sailor._base.fetch fetch_data 52 [ERROR] test
```

So the log origin points to the module, function and line where `LOG.error('test')` actually is

I only get this same result with `log_with_warning` if I set `stacklevel=3` for self.log (so forget my +1 thing from above)
`log_adapter.log_with_warning(DataNotFoundWarning())`

```
sailor._base.fetch fetch_data 53 [WARNING] No data found for given parameters.
```

If I don't modify stacklevel I get
```
sailor._base.fetch log_with_warning 32 [WARNING] No data found for given parameters.
```

which points to the correct module, but the function and line number is from the utils module which is weird (and not helping).

Reply:
I agree with Oliver, that the function and line number in the logs should not be misleading, especially if `stacklevel` is not specified. I also get only the expected results for `stacklevel = 3`. So I assume the correct offset is +2 and not +1. The same seems to be true for the warning_stacklevel (+2 is also needed). What about adding the following code to the `WarningAdapter` (mostly copied from `logging.LoggingAdapter`, changes are the first two code lines within the method):
```
def log(self, level, msg, *args, **kwargs):
    """
    Delegate a log call to the underlying logger, after adding
    contextual information from this adapter instance.
    """
    stacklevel_offset = 2
    kwargs["stacklevel"] = kwargs.get("stacklevel", 1) + stacklevel_offset
    if self.isEnabledFor(level):
        msg, kwargs = self.process(msg, kwargs)
        self.logger.log(level, msg, *args, **kwargs)
```
As far as I can see this will not only resolve the problem when `log_with_warning` is called, but calls like `adapter.error`  will also use the correct stacklevel.

Reply:
I made the changes as we agreed on:
- `WarningAdapter` increases the stacklevel-value as needed (done by overwriting `.log()`)
- `log_with_warning()` does not take the stacklevel parameter at all (of course it can still be passed to `.warning()` or `.error()`)